### PR TITLE
[3651] Improve isDescendantOf used in diagram

### LINKED ISSUE #3651: Improve isDescendantOf method
When dragging a node in a large diagram the method `isDescendantOf `is really slow.
On this graph representing 7second of computing time, the method is taking around 33% of the time.
image

This is because we're looking for the nodes needed to find the correct descendant during every call of `isDescendantOf`, we could use a map to retrieve the needed node faster

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3651

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts [hunk 1]
```diff
@@ -133,11 +131,11 @@ const getParameters: GetParameters = (movingNode, nodeA, nodeB, visiblesNodes, l
       y: (movingNode.positionAbsolute?.y ?? 0) + (nodeB.height ?? 0) / 2,
     };
   } else {
-    centerB = getNodeCenter(nodeB, visiblesNodes);
+    centerB = getNodeCenter(nodeB, nodeInternals);
   }
   const horizontalDifference = Math.abs(centerA.x - centerB.x);
   const verticalDifference = Math.abs(centerA.y - centerB.y);
-  const isDescendant = isDescendantOf(nodeB, nodeA, (nodeId) => visiblesNodes.find((node) => node.id === nodeId));
+  const isDescendant = isDescendantOf(nodeB, nodeA, nodeInternals);
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts:138]: `nodeInternals` contains *all* the nodes, including hidden ones, right?
If that's the case, doesn't this change the behavior? Before, we only looked into visible nodes. It doesn't seem like `isDescendantOf` tests for visiblity.

**Review by pcdavid — CHANGES_REQUESTED** [review]
Not sure if there is a possible regression with hidden nodes, but otherwise I confirm the performance improvements:

Before:

https://github.com/eclipse-sirius/sirius-web/assets/10608/c4f5af9f-5710-41e2-b5f7-44ce5a796fc9

After:

https://github.com/eclipse-sirius/sirius-web/assets/10608/7c52551c-d25d-471f-9ab8-cc8e71222cda

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts (hunk 1 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts:138]: Thank you, I missed that indeed

Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts:138]: It seems that the variable was called `visiblesNodes` but we were passing all nodes when calling it

**Review by pcdavid — APPROVED** [review]

**Comment by sbegaudeau** [comment]
I've not read a single line of this code, but since the PR has been approved, I've just resolved the conflict in the CHANGELOG since new commits have been merged on master and I'll merge it.

**Pull request merged**